simple_transformers/modality_processors.py

```diff
@@ -0,0 +1,301 @@
+        position_embeddings = position_embeddings.expand(batch_size, seq_len, -1)
+
+        # Scale input embeddings by sqrt of d_model. Unclear why, but seems to be standard practice
+        # https://datascience.stackexchange.com/questions/87906/transformer-model-why-are-word-embeddings-scaled-before-adding-positional-encod/87909#87909
+        embeddings = (input_embeds * math.sqrt(self.config.d_model)) + position_embeddings
+        embeddings = self.LayerNorm(embeddings)
+        embeddings = self.dropout(embeddings)
+        return embeddings, 1 - tokens['attention_mask']
+
+    def output_embeddings_to_logits(self, embs):
+        logits = embs @ th.transpose(self.word_embeddings.weight, 0, 1)
+        return logits
+
+    def logits_to_output(self, logits):
+        _, token_ids = th.max(logits, dim=-1)
+        return self.tokenizer.batch_decode(token_ids)
+
+
+class ImageProcessor(Processor):
+    """
+    Processes PIL.Images or np.arrays images into input embeddings. Uses a huggingface pretrained image preprocessor.
+    Assumes all images are the same size (if np.array) or made to be the same size using the preprocessor (PIL.Image).
+    Since images are all the same size, forward pass creates a constant sized attention mask.
+    REQUIRES: 'image_size', 'patch_size', and 'num_channels', kwargs
+    """
+    def __init__(self, config: SimpleNamespace, **kwargs):
+        super().__init__(config)
+        # Patching logic taken from : https://github.com/huggingface/transformers/blob/v4.27.1/src/transformers/models/vit/modeling_vit.py#L141
+        self.image_size = kwargs['image_size']
+        self.patch_size = kwargs['patch_size']
+        self.num_patches = (self.image_size[0] // self.patch_size[0]) * (self.image_size[1] // self.patch_size[1])
+
+        self.num_channels = kwargs['num_channels']
+        self.img_processor = CLIPImageProcessor.from_pretrained("openai/clip-vit-base-patch32")
+        self.patch_embeddings = nn.Conv2d(kwargs['num_channels'], config.d_model,
+                                          kernel_size=self.patch_size, stride=self.patch_size)
+        self.position_embeddings = nn.Embedding(self.num_patches + 1, config.d_model)  # +1 for cls token
+        self.cls_embedding = nn.Parameter(th.randn(config.d_model))
+        position_ids = th.arange(start=0, end=self.num_patches + 1, step=1)
+        # register buffer (these are constant tokens and NOT token embeddings)
+        self.register_buffer('position_ids', position_ids)
+        self.to(self.config.device)
+
+    def forward(self, images: Union[List[Image], np.array], att_mask: None) -> Tuple[th.Tensor, th.Tensor]:
+        """
+        :param images: If using PIL.Images, then preprocess using clip image preprocessing. This includes normalization
+                       and resizing to image size 224. Otherwise, condition directly on provided np.array
+        """
+        if type(images) == np.ndarray:
```

Comment:
```suggestion
        if isinstance(images, np.ndarray)
```
In general type checking should be done with `isinstance` since it supports inheritance.